File: java-base/src/main/java/net/cryptic_game/backend/base/netty/EventLoopGroupHandler.java
package net.cryptic_game.backend.base.netty;

import io.netty.channel.EventLoopGroup;
import io.netty.channel.epoll.Epoll;
import io.netty.channel.epoll.EpollEventLoopGroup;
import io.netty.channel.nio.NioEventLoopGroup;
import lombok.Data;

@Data
public class EventLoopGroupHandler {

    private static final boolean EPOLL = Epoll.isAvailable();

    private final EventLoopGroup bossGroup;
    private final EventLoopGroup workGroup;

    public EventLoopGroupHandler() {
        this.bossGroup = EPOLL ? new EpollEventLoopGroup(1) : new NioEventLoopGroup(1);
        this.workGroup = EPOLL ? new EpollEventLoopGroup() : new NioEventLoopGroup();
    }

    private void shutdown() {
        if (!this.bossGroup.isShutdown()) this.bossGroup.shutdownGracefully();
        if (!this.workGroup.isShutdown()) this.workGroup.shutdownGracefully();
    }
}

File: java-base/src/main/kotlin/net/cryptic_game/backend/base/netty/EventLoopGroupHandler.kt
package net.cryptic_game.backend.base.netty

import io.netty.channel.EventLoopGroup
import io.netty.channel.epoll.Epoll
import io.netty.channel.epoll.EpollEventLoopGroup
import io.netty.channel.nio.NioEventLoopGroup

class EventLoopGroupHandler {
    val bossGroup: EventLoopGroup = if (EPOLL) EpollEventLoopGroup(1) else NioEventLoopGroup(1)
    val workGroup: EventLoopGroup = if (EPOLL) EpollEventLoopGroup() else NioEventLoopGroup()

    private fun shutdown() {
        if (!bossGroup.isShutdown) bossGroup.shutdownGracefully()
        if (!workGroup.isShutdown) workGroup.shutdownGracefully()
    }

    companion object {
        private val EPOLL = Epoll.isAvailable()
    }

}
